File: programmers/박소민/lv1/유연근무제.js
/*
출근 희망 시각 + 10분 (주말 제외)
매일 한 번씩만 출근
모든 시각은 시에 100을 곱하고 분을 더한 정수

-> 일주일 동안 늦지 않고 출근

설계
1. 시간복잡도 o(n) 방식으로 설계. timelogs를 한 번만 순회
직원 수만큼의 boolean 배열
schedules를 미리 출근 시간으로 변경
forEach 분 + 10
엣지 케이스 (필요할까?)
1) 분 + 10 >= 60인 경우 -> 시 + 1, 분 - 60
2) 시 >= 24인 경우 -> 시 - 24

23시 48분이 출근시간
23시 58분 마감시간
00시 00분 출근이라면...

2. 주말인 경우(6, 7) continue
2.1. 아닌 경우 내부 배열을 순회하면서 확인

3. true 값의 갯수 반환
*/

function solution(schedules, timelogs, startday) {
  const n = schedules.length;
  const deadline = schedules.map((v) => {
    let hour = parseInt(v / 100);
    let minutes = (v % 100) + 10;

    if (minutes >= 60) {
      hour++;
      minutes -= 60;
    }

    // 나중에
    if (hour >= 24) {
      hour -= 24;
    }

    return hour * 100 + minutes;
  });

  const onTime = Array(n).fill(true);

  for (let worker = 0; worker < n; worker++) {
    let currentDay = startday;

    for (const time of timelogs[worker]) {
      const isWeekend = currentDay % 7 === 0 || currentDay % 7 === 6;
      if (isWeekend) {
        currentDay++;
        continue;
      }

      if (!onTime[worker]) continue;

      if (time > deadline[worker]) {
        onTime[worker] = false;
      }

      currentDay++;
    }
  }

  return onTime.filter((v) => v === true).length;
}

File: programmers/박소민/lv1/유연근무제.py
# 파이썬 dict: 키의 값(내용물)을 해시 ➡️ 인덱스 ➡️ 저장 (값 불변 필수!) (키 참조 주소, 값 참조 주소)
# JS {}: 무조건 키를 문자열로 강제 변환 ➡️ 문자열을 해시 ➡️ 인덱스 ➡️ 저장
# JS Map: 객체의 경우 메모리 주소(참조값) 자체를 해시 ➡️ 인덱스 ➡️ 저장 (값 변해도 주소만 같으면 됨!)
# 최악의 경우 (O(n)): 테이블이 2/3 차서 리사이징(이사) 할 때, 또는 해시 충돌이 꼬리에 꼬리를 물 때.

def solution(schedules, timelogs, startday):
    answer = 0
    
    for i, arr in enumerate(timelogs):
        day = startday
        time = schedules[i] + 10
        
        if time % 100 >= 60:
            time += 40
        
        # for else 좋다.
        for val in arr:
            # if in 도 좋다
            if day % 7 in (0, 6):
                day += 1
                continue
                
            if val > time:
                break
            
            day += 1
                
        else:
            answer += 1
        
    return answer